Dev/motor drivers

jsduenass opened the pull request:
create a ROS node that handles the dual max 14870  motor driver
- include a custom version of the dual max 14870   library by pololu
- creates launch file intended to be used to run the robot 
- set up ROS parameters to use multiple instances of the same node. Setting the GPIO pins connected to the raspberry.

jsduenass committed:
[FEAT]: add driver library

jsduenass committed:
docs: raspberry pin out

jsduenass committed:
break experomental

jsduenass committed:
remove submodule

jsduenass committed:
[FEAT]: add library

jsduenass committed:
add setup

jsduenass committed:
Merge branch 'main' into dev/motor-drivers

jsduenass committed:
[FIX]: driver lib and docker file
- driver lib variable typo
- docker file clean unnecessary  configurations

jsduenass committed:
[FEAT]: driver node
- create driver node subscriber
- document raspberry set up process

jsduenass committed:
[REFACT]: motor_driver subs
- split call_back for indivual motor velocity message
- doc  testing pub messages

jsduenass committed:
[FEAT]: driver pin out param
- driver pinout configured through ros param

jsduenass committed:
[FEAT]: driver pin out param
- driver pinout configured through ros param
- manage exceptions of pinout param not being found

jsduenass commented:
@cychitivav @Canborda 
ready for review!!!

cychitivav commented on Dockerfile:
```diff
@@ -11,6 +11,7 @@ RUN apt-get update && \
 
 RUN source /opt/ros/noetic/setup.bash
 
+
```
I do not see necessary this space
```suggestion
```

cychitivav commented on LICENSE:
I prefer to use this library like a submodule or create a new

jsduenass replied:
Alright I will fork the library repo and then added here as a submodule

cychitivav commented on README.md:
We are not using submodules yet

cychitivav commented on config/drivers_pinout.yaml:
why use one file with the parameters for the two drivers and then two other files, one for each driver?

jsduenass replied:
I was still testing and didn't know which approach was better to use a config file for both motor or separate each in its config file. I decided the latter so I remove this file

cychitivav commented on docs/source/imgs/raspberry-pi-3-PWM-pins.png:
I prefer to do not upload images.

cychitivav commented on lib/max14870-driver/LICENSE.txt:
License file duplicated

jsduenass replied:
This would still hold if we use the submodule

cychitivav replied:
and why not create our own code?

cychitivav commented on docs/source/index.md:
```diff
@@ -1,2 +1,26 @@
+
+
+## Raspbery pin out 
+![rpi pin out](./imgs/raspberry-pi-3-PWM-pins.png)
```
I would like to use this image
![image](https://github.com/cychitivav/pai/assets/30636259/018453a2-3278-494c-a706-0de937c5933a)
```suggestion
![rpi pin out](https://github.com/cychitivav/pai/assets/30636259/018453a2-3278-494c-a706-0de937c5933a)
```

cychitivav commented on launch/run_robot.launch:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0"?>
+<launch>
+  
+     <rosparam file="$(find pai)/config/drivers.yaml" command="load" />
```
This cannot be loaded, and I think it is not necessary.
```suggestion
```

cychitivav commented on launch/run_robot.launch:
```diff
@@ -0,0 +1,38 @@
+    />
+    
+    <!-- Right differential driver  -->
+    <arg name="driver_R"  command="load"   file="$(find pai)/config/driver_R_pinout.yaml" />
+    <node   name="right_driver" 
+            pkg="pai" output="screen"
+            type="motor_driver.py" 
+        >
+        <param name="driver_pinout"              value="$(arg driver_R)"/>     
+        <remap from="/wheel_A_ctrl/command" to="/front_right_wheel_ctrl/command"/>
+        <remap from="/wheel_B_ctrl/command" to="/back_right_wheel_ctrl/command"/>
+    </node>
+
+    
+    <!-- Left differential driver  -->
+    <arg name="driver_L"  command="load"   file="$(find pai)/config/driver_R_pinout.yaml" />
+    <node   name="left_driver" 
+            pkg="pai" output="screen"
+            type="motor_driver.py"     
+     >
+        <param name="driver_pinout"              value="$(arg driver_L)"/>         
+        <remap from="/wheel_A_ctrl/command" to="/front_left_wheel_ctrl/command"/>
+        <remap from="/wheel_B_ctrl/command" to="/back_left_wheel_ctrl/command"/>    
+    </node>
```
Arguments cannot be loaded with `command="load"`

```suggestion
    <rosparam file="$(find pai)/config/driver_R_pinout.yaml" command="load" />
    <node   name="right_driver" 
            pkg="pai" output="screen"
            type="motor_driver.py" 
        >
        <param name="driver_pinout"              value="$(arg driver_R)"/>     
        <remap from="/wheel_A_ctrl/command" to="/front_right_wheel_ctrl/command"/>
        <remap from="/wheel_B_ctrl/command" to="/back_right_wheel_ctrl/command"/>
    </node>

    
    <!-- Left differential driver  -->
    <rosparam file="$(find pai)/config/driver_L_pinout.yaml" command="load" />
    <node   name="left_driver" 
            pkg="pai" output="screen"
            type="motor_driver.py"     
     >
        <param name="driver_pinout"              value="$(arg driver_L)"/>         
        <remap from="/wheel_A_ctrl/command" to="/front_left_wheel_ctrl/command"/>
        <remap from="/wheel_B_ctrl/command" to="/back_left_wheel_ctrl/command"/>    
    </node>
```

cychitivav commented on scripts/motor_driver.py:
```diff
@@ -0,0 +1,60 @@
+
+from dual_max14870_rpi import DualMotorDriver
+
+try:
+    pin = rospy.get_param('driver_pinout')
+    rospy.logwarn("no 'driver_pinout' param found using default config file")
+except:
+    try:
+        (pin, namespace) =  rosparam.load_file("./config/driver_L_pinout.yaml", )
+        rospy.logwarn("no default config file using hard coded pinout")
+    except:
+        pin ={ 'FAULT':23, 'PWM2':24, 'PWM1':12, 'DIR2':16, 'DIR1':20, 'EN': 21}
```
This is a bit exaggerated, better to pass the default parameters and get the parameters using as namespace the node's name.
```suggestion
pin = rospy.get_param(rospy.get_name(), {'FAULT':23, 'PWM2':24, 'PWM1':12, 'DIR2':16, 'DIR1':20, 'EN': 21})
```

cychitivav commented on scripts/motor_driver.py:
```diff
@@ -0,0 +1,60 @@
+
+
+if __name__ == '__main__':
+    rospy.init_node("Computer", anonymous=True)
```
A little suggestion
```suggestion
    rospy.init_node("Driver", anonymous=True)
```

cychitivav commented on scripts/motor_driver.py:
As I said before, I prefer not to use the dual_max14870_rpi library.

jsduenass committed:
Dockerfile remove sourcing ROS

Co-authored-by: Cristian Chitiva <cychitivav@unal.edu.co>

jsduenass committed:
streamline getting rosparameters

Co-authored-by: Cristian Chitiva <cychitivav@unal.edu.co>

jsduenass committed:
Update scripts/motor_driver.py

Co-authored-by: Cristian Chitiva <cychitivav@unal.edu.co>

jsduenass committed:
Update launch/run_robot.launch

Co-authored-by: Cristian Chitiva <cychitivav@unal.edu.co>

jsduenass committed:
[FIX]: rosparam in launch file

Co-authored-by: Cristian Chitiva <cychitivav@unal.edu.co>

jsduenass committed:
change image

Co-authored-by: Cristian Chitiva <cychitivav@unal.edu.co>

jsduenass committed:
[DOCS]: separate raspberry pin documentation
-  split electronics documentation into its own page
- remove local image

jsduenass committed:
[FEAT]: remove max14870 lib from source code
- remove the lib and add it as a submodule

jsduenass committed:
[FIX]: remove duplicate pionut documentation
- remove pionut from index file in documentation as is duplicated
- remove redundant drivers_pinout

jsduenass commented:
alright @cychitivav  how do you find the changes?

cychitivav commented:
In general lgtm, but I prefer to create our own code, I see that the `Driver` class inherits from `DualMotorDriver` and I think they are the same thing. This would go against the OOP paradigm.

jsduenass commented:
It is true that the Driver class is not doing that much is just working as a wrapper to communicate with the ROS messages. And the library used is not that complex. However, I think that the implementation of this change would better be handle in a separate issue and merge what we have so far.

cychitivav commented on launch/run_robot.launch:
```diff
@@ -0,0 +1,37 @@
+    />
+    
+    <!-- Right differential driver  -->
+    <rosparam file="$(find pai)/config/driver_R_pinout.yaml" command="load" />
+    <node   name="right_driver" 
+            pkg="pai" output="screen"
+            type="motor_driver.py" 
+        >
+        <param name="driver_pinout"              value="$(arg driver_R)"/>     
+        <remap from="/wheel_A_ctrl/command" to="/front_right_wheel_ctrl/command"/>
+        <remap from="/wheel_B_ctrl/command" to="/back_right_wheel_ctrl/command"/>
+    </node>
+
+    
+    <!-- Left differential driver  -->
+    <rosparam file="$(find pai)/config/driver_L_pinout.yaml" command="load" />
+    <node   name="left_driver" 
+            pkg="pai" output="screen"
+            type="motor_driver.py"     
+     >
+        <param name="driver_pinout"              value="$(arg driver_L)"/>         
```
It is also necessary to remove the `args`.
```suggestion
    <rosparam file="$(find pai)/config/driver_R_pinout.yaml" command="load" />
    <node   name="right_driver" 
            pkg="pai" output="screen"
            type="motor_driver.py" 
        >
        <remap from="/wheel_A_ctrl/command" to="/front_right_wheel_ctrl/command"/>
        <remap from="/wheel_B_ctrl/command" to="/back_right_wheel_ctrl/command"/>
    </node>

    
    <!-- Left differential driver  -->
    <rosparam file="$(find pai)/config/driver_L_pinout.yaml" command="load" />
    <node   name="left_driver" 
            pkg="pai" output="screen"
            type="motor_driver.py"     
     >   
```

jsduenass replied:
I think with this change both nodes will be left with the parameter of the last definition. in this case `driver_L_pinout.yaml`.
but let's include the change to see how it goes.

cychitivav commented:
> It is true that the Driver class is not doing that much is just working as a wrapper to communicate with the ROS messages. And the library used is not that complex. However, I think that the implementation of this change would better be handle in a separate issue and merge what we have so far.

Ok, I think you are right, for now let's fix the last changes and in another PR or issue we will discuss about the submodule.

jsduenass commented:
> Ok, I think you are right, for now let's fix the last changes and in another PR or issue we will discuss about the submodule

Sounds good

jsduenass committed:
[FIX]: launch file param

Co-authored-by: Cristian Chitiva <cychitivav@unal.edu.co>
